coda_v2_python_client/firebase_client_wrapper.py

```diff
@@ -146,12 +151,18 @@ def create_next_segment(self, dataset_id, transaction=None):
         log.debug(f"Creating next dataset segment with id {next_segment_id}")
 
         code_schemes = self.get_all_code_schemes(current_segment_id, transaction=transaction)
-        self.add_and_update_code_schemes(next_segment_id, code_schemes)
-
         users = self.get_user_ids(current_segment_id, transaction=transaction)
-        self.set_user_ids(next_segment_id, users)
 
-        self.set_segment_count(dataset_id, next_segment_count)
+        # Note: Since a transaction can fail if it contains read operations after write operations, we
+        # pass `segment count` to set_user_ids to prevent read operation of "segment count" after the write
+        # operation performed by add_and_update_code_schemes. In addition it reduces the number of read
+        # operations done to acquire segment count.
+        self.add_and_update_code_schemes(next_segment_id, code_schemes, transaction=transaction, segment_count=next_segment_count)
+        self.set_user_ids(next_segment_id, users, transaction=transaction, segment_count=next_segment_count)
+        self.set_segment_count(dataset_id, next_segment_count, transaction=transaction)
```

Comment:
I think from your comment you could tell that this didn't quite feel clean.

I think the better way to do this is to have 2 functions, set_dataset_user_ids and set_segment_user_ids. Then you call set_segment_user_ids, which doesn't need a read, here

Reply:
Thanks updated. I've not yet renamed `set_user_ids` to `set_dataset_user_ids`, I will open another PR for that since there are other functions that need to be renamed to include `dataset`.